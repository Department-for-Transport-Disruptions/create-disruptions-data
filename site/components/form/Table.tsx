import { ReactElement, ReactNode } from "react";

export interface TableStyles {
    width?: string;
}

export interface CellProps {
    value: string | ReactNode;
    styles?: TableStyles;
}

interface TableProps {
    caption?: { text: string; size: "s" | "m" | "l" };
    columns?: string[];
    rows: { header?: string | ReactNode; cells: string[] | ReactNode[] | CellProps[] }[];
}

const isCellProps = (cell: string | ReactNode | CellProps): cell is CellProps => {
    if (cell && typeof cell === "object" && "value" in cell) {
        return true;
    } else {
        return false;
    }
};

const Table = ({ caption, columns = [], rows }: TableProps): ReactElement => {
    return (
        <table className="govuk-table">
            {caption ? (
                <caption className={`govuk-table__caption govuk-table__caption--${caption.size}`}>
                    {caption.text}
                </caption>
            ) : (
                ""
            )}
            <thead className="govuk-table__head">
                <tr className="govuk-table__row">
                    {columns.map((column) => (
                        <th scope="col" className="govuk-table__header align-middle px-2" key={column}>
                            {column}
                        </th>
                    ))}
                </tr>
            </thead>
            <tbody className="govuk-table__body">
                {rows.map((row, index) => (
                    <tr className="govuk-table__row" key={`row-${index}`}>
                        {row.header && (
                            <th scope="row" className="govuk-table__header align-middle">
                                {row.header}
                            </th>
                        )}
                        {row.cells.map((cell, i) => (
                            <td
                                className={`govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl ${
                                    isCellProps(cell) && cell.styles?.width ? cell.styles?.width : ""
                                }`}
                                key={i}
                            >
                                {isCellProps(cell) ? cell.value : cell}
                            </td>
                        ))}
                    </tr>
                ))}
            </tbody>
        </table>
    );
};

export default Table;
