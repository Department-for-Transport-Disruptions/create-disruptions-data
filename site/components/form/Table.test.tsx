import renderer from "react-test-renderer";
import { describe, it, expect } from "vitest";
import Table from "./Table";

describe("Table", () => {
    it("should render correctly", () => {
        const tree = renderer
            .create(
                <Table
                    caption={{ text: "test table", size: "l" }}
                    columns={["test col 1", "test col 2"]}
                    rows={[
                        {
                            header: "test row 1",
                            cells: ["1"],
                        },
                        {
                            header: "test row 2",
                            cells: ["2"],
                        },
                    ]}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("should render correctly without a header", () => {
        const tree = renderer
            .create(
                <Table
                    caption={{ text: "test table", size: "l" }}
                    columns={["test col 1", "test col 2"]}
                    rows={[
                        {
                            cells: ["1"],
                        },
                        {
                            cells: ["2"],
                        },
                    ]}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("should render correctly with jsx element", () => {
        const tree = renderer
            .create(
                <Table
                    caption={{ text: "test table", size: "l" }}
                    columns={["test col 1", "test col 2"]}
                    rows={[
                        {
                            header: "test row 1",
                            cells: [<p key="test">test 1</p>],
                        },
                        {
                            header: "test row 2",
                            cells: ["2"],
                        },
                    ]}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
