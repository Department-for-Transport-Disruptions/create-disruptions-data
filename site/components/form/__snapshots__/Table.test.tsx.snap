// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > should render correctly 1`] = `
<table
  className="govuk-table"
>
  <caption
    className="govuk-table__caption govuk-table__caption--l"
  >
    test table
  </caption>
  <thead
    className="govuk-table__head"
  >
    <tr
      className="govuk-table__row"
    >
      <th
        className="govuk-table__header align-middle"
        scope="col"
      >
        test col 1
      </th>
      <th
        className="govuk-table__header align-middle"
        scope="col"
      >
        test col 2
      </th>
    </tr>
  </thead>
  <tbody
    className="govuk-table__body"
  >
    <tr
      className="govuk-table__row"
    >
      <th
        className="govuk-table__header align-middle"
        scope="row"
      >
        test row 1
      </th>
      <td
        className="govuk-table__cell align-middle"
      >
        1
      </td>
    </tr>
    <tr
      className="govuk-table__row"
    >
      <th
        className="govuk-table__header align-middle"
        scope="row"
      >
        test row 2
      </th>
      <td
        className="govuk-table__cell align-middle"
      >
        2
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table > should render correctly with jsx element 1`] = `
<table
  className="govuk-table"
>
  <caption
    className="govuk-table__caption govuk-table__caption--l"
  >
    test table
  </caption>
  <thead
    className="govuk-table__head"
  >
    <tr
      className="govuk-table__row"
    >
      <th
        className="govuk-table__header align-middle"
        scope="col"
      >
        test col 1
      </th>
      <th
        className="govuk-table__header align-middle"
        scope="col"
      >
        test col 2
      </th>
    </tr>
  </thead>
  <tbody
    className="govuk-table__body"
  >
    <tr
      className="govuk-table__row"
    >
      <th
        className="govuk-table__header align-middle"
        scope="row"
      >
        test row 1
      </th>
      <td
        className="govuk-table__cell align-middle"
      >
        <p>
          test 1
        </p>
      </td>
    </tr>
    <tr
      className="govuk-table__row"
    >
      <th
        className="govuk-table__header align-middle"
        scope="row"
      >
        test row 2
      </th>
      <td
        className="govuk-table__cell align-middle"
      >
        2
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table > should render correctly without a header 1`] = `
<table
  className="govuk-table"
>
  <caption
    className="govuk-table__caption govuk-table__caption--l"
  >
    test table
  </caption>
  <thead
    className="govuk-table__head"
  >
    <tr
      className="govuk-table__row"
    >
      <th
        className="govuk-table__header align-middle"
        scope="col"
      >
        test col 1
      </th>
      <th
        className="govuk-table__header align-middle"
        scope="col"
      >
        test col 2
      </th>
    </tr>
  </thead>
  <tbody
    className="govuk-table__body"
  >
    <tr
      className="govuk-table__row"
    >
      <td
        className="govuk-table__cell align-middle"
      >
        1
      </td>
    </tr>
    <tr
      className="govuk-table__row"
    >
      <td
        className="govuk-table__cell align-middle"
      >
        2
      </td>
    </tr>
  </tbody>
</table>
`;
